Port ZXing Android to Camera2

Denrage opened the pull request:
During investigation of adding new features to this library we noticed that it's still using the obsolet Android.Hardware.Camera library. To have it future proof i refactored the code of the CameraController.android to use the Camera2 API now. This is just a port and it's not removing functionality or adding any.
The only hiccup i had was the FastJavaByteArray. Because the ImageReader is not giving out a pointer on receiving an image, the FastJavaByteArray can't be used anymore (or it would slow it down)

Commit:
Ported to camera2

Commit:
Refactoring

Commit:
Fixed distorted preview

Commit:
Cleanup

Redth commented:
This is great! thanks for the work put into this! I'm going to get an update out first that fixes some issues and still allows for < android api 21 support and then will merge this after.

The only issue I'm seeing so far on this PR is that the orientation seems off 90 degrees... The sample which doesn't auto rotate doesn't find 1D barcodes unless i rotate the device, so perhaps the buffer needs a quick rotate to be in the expected orientation...

jameswestgate commented:
I just want to say this is an awesome pull request but it seemed to break Autofocus for me and I couldn’t get scanning to work at all on Pixel 5 running Android Q. Other scenarios worked well though.

Denrage committed:
Removed obsolet boundary check for auto focus

Denrage commented:
> I just want to say this is an awesome pull request but it seemed to break Autofocus for me and I couldn’t get qr code scanning to work at all on Pixel 5 running Android Q. Other scenarios worked well though.

At least in the sample autofocus is working perfectly on my Huawei Mate 10 Pro. The Touch event to force a focus is also working. One caveat: the touch event can only be handled if there is no overlay bc currently the overlay is not propagating the touch event down to the surface view

Denrage commented:
> This is great! thanks for the work put into this! I'm going to get an update out first that fixes some issues and still allows for < android api 21 support and then will merge this after.
> 
> The only issue I'm seeing so far on this PR is that the orientation seems off 90 degrees... The sample which doesn't auto rotate doesn't find 1D barcodes unless i rotate the device, so perhaps the buffer needs a quick rotate to be in the expected orientation...

Do you have an example barcode which is not working? Bc for me it's working however the bar code is orientated or the phone is orientated. I chose a random barcode from the internet for that test
https://cdn.discordapp.com/attachments/396196956830957571/801015315843776522/SVID_20210119_100809_1.mp4

jameswestgate commented:
> Need to test it in simulator for Pixel 5, rlly have no idea why it wouldn't work on that

Thank you for responding! I will put together a minimum example for you in the next few days, to validate that it's not something else.

jameswestgate commented:
I ran the camera2 branch unmodified on my physical Pixel5 and it did not scan a QR code. I then got my Galaxy s9 and it **did** scan. The s9 is running Android 9. However, the latest master from @Redth works fine on Pixel 5.

I will debug and see if I can find out what's happening.

Denrage commented:
> I will debug and see if I can find out what's happening.
That would be great!

The only thing i can imagine is that the pixel has an android which is too old for camera2. Please tell me any finding you will have there

jameswestgate commented:
This is the latest Pixel5 running Android 11. 
I have looked into this further, specifically `CameraAnalyser.DecodeFrame`

On @Redth 's master branch (has some updated commits), the frame is being decoded as follows:

> Decode Time: 80.6448 ms (width: 800, height: 600, degrees: 90, rotate: True)

On your branch

> Decode Time: 92.926 ms (width: 352, height: 288)

There is a difference in how the width and height for decoding are calculated in these two codebases, and I think this is where the problem lies. Can I suggest maybe you open Issues on your branch of the code and we can collaborate further there and resolve the issue outside of this pull request? 

It could also be helpful to merge the latest commits from @Redth 's master into this branch?

All the best, James

Redth commented:
I wonder if there's some dpi calculation missing?

jameswestgate commented:
I'll look into how the two repos work these values out and see if that has something to do with it

jameswestgate commented:
Ok good news. I have got QR scanning working on the Pixel5 with camera2. 

A size is calculated in `CameraController.GetOptimalSize`, this tries to use the best aspect ratio it can find from all available sizes, however, this results in the very low width and height being selected as it is the best ratio. There's equivalent code in the master repo that sets a constraint around minimum widths and heights, introducing it returns the same dimensions as master. I also notice that this code doesn't call `MobileBarcodeScanningOptions.GetResolution` as it does in master, to allow developers to set their preferred resolutions.

in `CameraAnalyser.DecodeFrame` for the Pixel5 to scan a QR code, it also needs the `options.TryHarder` value set, this slows scan time down quite a bit (+- 700ms), so I don't know if we can provide better data to the `barcodeReader`.  A `barcodeReader` was being created every frame, I have cached this to improve performance.

I will add a pull request to the camera2 branch so this can be tested further. J

jameswestgate commented:
Also note, the `barcodeReader` has AutoRotate set to true, so I wonder why the barcode can't be read at any orientation ... maybe to do with width and height supplied.

jameswestgate commented:
The reason for the width and height is because this is calculated to be the best aspect ratio although they are quite small values. I am tryng to see how this relates to the data provided to the `barcodeReader` as well as `PlanarYUVLuminanceSource` which seems to require the width and height of the data.

jameswestgate commented:
Well, this was quite a rabbit hole to go down, but I have this working on my Pixel 5.

I adjusted the algorithm to return the same size as in the master branch (by excluding smaller sizes), I looked at the ImageReader because no matter what I tried, the decoder did not want to find a QR code.

When outputting the yuv bytes to disk as a jpeg, it become apparent there was a yuv conversion issue on the Pixel5. 
It turned out the code in `CameraEventsListener.ImageToByteArray` may not always work correctly, and I replaced it with code converted from this answer: 

https://stackoverflow.com/questions/52726002/camera2-captured-picture-conversion-from-yuv-420-888-to-nv21. 

Now I know what the issue is, I will see if there is a better way of getting a LuminanceSource from the preview image, if not I will optimise and provide the code as a pull request to the camera2 branch.

jameswestgate committed:
Add new Nv21  implementation

jameswestgate committed:
Dont create new BarcodeReader on every decode

jameswestgate commented:
At risk of pull request inception, see https://github.com/Denrage/ZXing.Net.Mobile/pull/1 for changes

Denrage commented:
So the main problem here was not the difference in size but the conversion of the YUV to Nv21 on the Pixel5? I still need to implement the forwarding of a desired size in regard of the options used (as you mentioned in an earlier comment). I want to merge master into the branch anyway in the end but will do that ofc too

jameswestgate commented:
I actually left the size in the end, because it was working fine and the performance was really good as a result. I think in the master branch when selecting the most appropriate size for an aspect ratio, it skips very small sizes because it is for the preview. 

> I want to merge master into the branch anyway in the end but will do that ofc too

The nuget package dependancy on FastAndroidCamera as well as [FastJavaArrayEx.android.cs](https://github.com/Denrage/ZXing.Net.Mobile/blob/master/ZXing.Net.Mobile/Android/FastJavaArrayEx.android.cs) and [FastJavaByteArrayYUVLuminanceSource.android.cs](https://github.com/Denrage/ZXing.Net.Mobile/blob/master/ZXing.Net.Mobile/Android/FastJavaByteArrayYUVLuminanceSource.android.cs) can also be removed

Denrage committed:
Merge pull request #1 from jameswestgate/android-camera2

Fix yuv to nv21

Denrage committed:
removed obsolet code

Denrage committed:
Removed FastAndroidCamera

Denrage committed:
Merge branch 'master' into android-camera2

# Conflicts:
#	ZXing.Net.Mobile/Android/CameraAccess/CameraAnalyzer.android.cs

Denrage committed:
Fixed Merge

Denrage committed:
rotate yuv output

Denrage commented:
- Merged YUV changes of jameswestgate
- rotated YUV output (saw it was 90° off in the jpeg debug)
- merged master and added the changes from master in the CameraAnalyzer
- Removed dependencies to the FastAndroidCamera

jameswestgate commented:
Tested this on my Pixel5. Great job cleaning it all up and merging to master.

Unfortunately the rotation code is showing the same symptoms of not fully understanding the yuv format output by the pixel5. Attached please find some images. 

**With rotate**
https://photos.app.goo.gl/7aBbMG2xn86DDkub9

**Without Rotate**
https://photos.app.goo.gl/43QL4UvJu3Hnwu3X8

**Without Rotate, Horizontal**
https://photos.app.goo.gl/wtQUdBkZgYCoUVns8

I also think a simple 90 degree rotate may be naive, to test this I held my phone horizontally. You can now see the image (without the rotation code) is 180 degree rotated. So without being able to look at the metadata of the image, or know the orientation of the phone, it is probably not possible to correct for this. I also wonder if this is necessary, since the BarcodeReader has AutoRotate set to true. 

I know @Redth has an issue with barcodes, I wonder if we can find a nice way of setting up a debug mode that we can keep in the code base so that we can get some images being passed to the BarcodeReader from him, or others, whilst we work through any bugs ...

Denrage commented:
Good to hear that it worked now on the pixel. Can't see the images (album is empty according to google), but yeah thinking over it a little the rotation is way to naive and needs to take the orientation of the phone into regard. I implemented it bc of the problem @Redth had with the rotation (even though 1D barcodes worked for me in any rotation of the phone). The question is, why the previous implementation had a rotation correction even though the reader is on autorotate

jameswestgate commented:
I have updated the photo links, hopefully they stay working :)

Cheesebaron commented:
https://user-images.githubusercontent.com/249719/108498269-e8c38080-72ac-11eb-8e25-428037440d68.mp4

Just tried this in one the Apps I work on. It seems to kind of work, but for every frame it analyzes it freezes the preview. Aspect looks slightly weird too.

Tested on OnePlus 8 and OnePlus 6.

Another issue is that `CameraResolutionSelector` is never called any more either.


EDIT: Seems like this was some issue with the NuGet package that we were consuming. Building this branch ourselves seems to have fixed the issues above.

keozx commented:
I found also Pixel 3a and 3XL does not seem to scan the codes 🤔

Denrage committed:
Fixed scanning on tablet and sped up conversion

Denrage committed:
Fixed distortion of camera preview

Denrage commented:
>  21-02-19-12-08-23.mp4 
> Just tried this in one the Apps I work on. It seems to kind of work, but for every frame it analyzes it freezes the preview. Aspect looks slightly weird too.
> 
> Tested on OnePlus 8 and OnePlus 6.
> 
> Another issue is that `CameraResolutionSelector` is never called any more either.
> 
> EDIT: Seems like this was some issue with the NuGet package that we were consuming. Building this branch ourselves seems to have fixed the issues above.

The aspect ratio was a thing that bugged me a long time but there were more important things to fix before that. Apparently you just need to give the holder.SetFixedSize a little bit of time to adjust. Shouldn't be distorted anymore

Commit:
Select lowest resolution for faster analyze

Cheesebaron commented:
We've been running this code in production for a while now. Without many issues. The main issue we've seen is this:

```
Java.Lang.RuntimeException: setParameters failed
  at Java.Interop.JniEnvironment+InstanceMethods.CallVoidMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x0006e] in <1fe1c2b415bf4a329521a6070272f9a8>:0
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x0002a] in <1fe1c2b415bf4a329521a6070272f9a8>:0
  at Android.Hardware.Camera.SetParameters (Android.Hardware.Camera+Parameters params) [0x00031] in <eb636a3761e1450594e38a59edb83b74>:0
  at ZXing.Mobile.CameraAccess.CameraController.ApplyCameraSettings () [0x0029b] in <c0c97e0a6fd341bb93224d28d1a94bbf>:0
  at ZXing.Mobile.CameraAccess.CameraController.RefreshCamera () [0x00009] in <c0c97e0a6fd341bb93224d28d1a94bbf>:0
  at ZXing.Mobile.CameraAccess.CameraAnalyzer.RefreshCamera () [0x00000] in <c0c97e0a6fd341bb93224d28d1a94bbf>:0
  at ZXing.Mobile.ZXingSurfaceView.SurfaceChanged (Android.Views.ISurfaceHolder holder, Android.Graphics.Format format, System.Int32 wx, System.Int32 hx) [0x00007] in <c0c97e0a6fd341bb93224d28d1a94bbf>:0
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) [0x00000] in <11d99176e2e94d8eaebfa4c45c1e46a4>:0
  at Android.App.SyncContext+<>c__DisplayClass2_0.<Post>b__0 () [0x00000] in <eb636a3761e1450594e38a59edb83b74>:0
  at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in <eb636a3761e1450594e38a59edb83b74>:0
  at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <eb636a3761e1450594e38a59edb83b74>:0
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.40(intptr,intptr)
  at java.lang.RuntimeException: setParameters failed
  at android.hardware.Camera.native_setParameters(Native Method)
  at android.hardware.Camera.setParameters(Camera.java:2138)
  at crc6480997b3ef81bf9b2.ZXingSurfaceView.n_surfaceChanged(Native Method)
  at crc6480997b3ef81bf9b2.ZXingSurfaceView.surfaceChanged(ZXingSurfaceView.java:84)
  at android.view.SurfaceView.updateSurface(SurfaceView.java:1300)
  at android.view.SurfaceView.setFrame(SurfaceView.java:646)
  at android.view.View.layout(View.java:24472)
  at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
  at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
  at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1103)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at androidx.constraintlayout.widget.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
  at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
  at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
  at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
  at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
  at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
  at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
  at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
  at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
  at com.android.internal.policy.DecorView.onLayout(DecorView.java:1225)
  at android.view.View.layout(View.java:24475)
  at android.view.ViewGroup.layout(ViewGroup.java:7383)
  at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:4260)
  at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3695)
  at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2618)
  at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9971)
  at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1010)
  at android.view.Choreographer.doCallbacks(Choreographer.java:809)
  at android.view.Choreographer.doFrame(Choreographer.java:744)
  at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:995)
  at android.os.Handler.handleCallback(Handler.java:938)
  at android.os.Handler.dispatchMessage(Handler.java:99)
  at android.os.Looper.loop(Looper.java:246)
  at android.app.ActivityThread.main(ActivityThread.java:8512)
  at java.lang.reflect.Method.invoke(Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

But that is for 7 out of 10k+ users.

Denrage commented:
Does this happen reproduceable? Because the error message is saying nothing at all and till now it didn't happen for me.
I'm rlly happy about that it is working so good

Cheesebaron commented:
@Denrage haven't seen it myself. It is a very rare crash, so I am not sure what exactly is going on.

Denrage commented:
Would be great if Android would be more specific with that error. Does it happen always on a certain device or is it actually random across different vendors and most of the time it works and sometimes it crashes? 
Idk if i should catch it and throw a custom exception for that because in either case the user of the lib needs to catch it at some point

Cheesebaron commented:
Looks like it is mainly on Android 11 devices:
![image](https://user-images.githubusercontent.com/249719/121433061-c6e52b00-c97b-11eb-8211-d04161e4cd8a.png)

Commit:
Use a minimum size to prevent using a resolution too small

keozx commented:
Hi, I came back to this and could test it further, this works very well in my testing, would be nice to resurrect this PR :)
The only thing I'm not sure is, we don't seem to need the `CameraResolutionSelectorDelegate` and company as the `GetOptimalPreviewSize` method does a good job selecting an optimal size, should this delegate be removed as it is not called anymore? Or may an adapted port of the delegate be of any use to make a custom size selection?

roubachof commented:
Please merge this :)

Cheesebaron commented:
@Redth anything holding this one back? I know it is a big change to review. However, is there anything that would help making this one easier for you to review?

Commit:
Updated ZXing Lib

Commit:
Add DebugHelper.android.cs class

Commit:
Improved read speed a little bit and better recognition for bar codes

Commit:
Spacings

Commit:
Allow tools folder

Commit:
Added ImageReceiver Tool

Commit:
typo fixed

Commit:
Prefere slight distortion but see whats on screen

DarkIrata commented:
Long story short, a customer found some problems scanning bar codes. I did improve it a little bit and changed a few more lines to make it a little bit faster (also for qr codes) but there is still some improvements to make which i will try to implement. Whil i don't have much hope it will be merged anytime soon.. at least there is a reason now to wait.  

Edit: Tests properly if you want to use it productive. The base functionallity worked fine for me in my tests, still..
Edit 2:  I will be on vacation and after that will try to fix and improve it further. Mainy having problems with "code 39" Barcodes.

Commit:
Use helper class CapturedImageData

Commit:
Call AutoFocus at least once

Commit:
if available and not other given, use continousvideo mode für auto focus

Commit:
Use ideal sizes and update accordingly

Commit:
Better LuminanceSource for nv21

Commit:
Small adjustment in yuv420 to nv21 converter

Commit:
Use the improved LuminanceSource with correct rotation

Commit:
Changed ImageReader instances to 3

Commit:
Some Debug Helper improvements

Commit:
Better initial image rotation

Commit:
Adjusted Log Message

Commit:
Fixed Bug where potrait images gets rotated because of device orientation

Commit:
Moved initializing orientationdata for logging

Commit:
Adjusted Optimal Sizes for quality and performance

Commit:
More samples since we got lower decode times

Commit:
Debug call example

Commit:
adjusted how the optimal sizes get determined

DarkIrata commented:
From this point i can't do much more. I need a few people to test on different devices. Especially tablets. Current state is:

- Reduced Resolutions from decode frames (Better results for barcodes, faster decode)
- Changed LuminanceSource to a Custom NV21 LuinanceSource for decoder with a proper rotation function
- Frames now get (if AutoRotate is false) automaticlly rotated based on Orientation and Sensor Rotation on initial as close as possible to what you see
- Added a Debugging Helper so you can kind of see what the decoder see
- Focus is not based on continous video initially 

Thing which could be improved are:
- (Theory) 2 ImageReader. One for the low resolution and one for a higher resolution. I don't know if it would be possible by OS restrictions, but in theory this would allow lower end devices get a proper quality picture. Since i don't have many devices to test, i don't know what the quality right now is
- Cropping, this would skyrock the decode time further.

Tested with:
- OnePlus 7 Pro  (~30-150ms Decode time)
- Huawei M20 (~10-130ms Decode time)

Commit:
Fallback for OptimalSize

Commit:
More OptimalSize adjustments

Commit:
Fixed image bending to a certain degree
